pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -377,13 +377,12 @@ func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) b
 		status.UpdateRevision != set.Status.UpdateRevision
 }
 
-// completeRollingUpdate completes a rolling update when all of set's replica Pods have been updated
+// completeUpdate completes an update when all of set's replica Pods have been updated
 // to the updateRevision. status's currentRevision is set to updateRevision and its' updateRevision
 // is set to the empty string. status's currentReplicas is set to updateReplicas and its updateReplicas
 // are set to 0.
-func completeRollingUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
-	if set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
-		status.UpdatedReplicas == status.Replicas &&
+func completeUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
```

Comment:
It works for the prescribed example, but after testing it manually I have  found a following problem:

```
pods: pod-0(rev1) pod-1(rev2) pod-2(rev3)

StatefulSet:
currentRevision: rev1
updateRevision: rev3
```
`$ kubectl delete pod-0`

```
pods: pod-0(rev3) pod-1(rev2) pod-2(rev3)

StatefulSet:
currentRevision: rev1
updateRevision: rev3
```

even though we have not completed the rev1 is not present anymore, and the current revision should be rev2 now

Reply:
As it stands the logic won't touch the currentRevision until everything is updated to the updateRevision.  @atiratree would the expectation be to set currentRevision to rev2 in this scenario?  I believe a RollingUpdate where the spec is updated partway through a rollout would have the same issue, but I can confirm this later today.

Reply:
RollingUpdate tries to update the newest first so there is always a pod present with the oldest revision, so it shouldn't be an issue here

Reply:
OK I've introduced a new function for calculating the current version that takes this into consideration.

Reply:
we should get additional opinion on this first, before polishing the code and testing

Reply:
I have also added a unit test for the new revision accounting.